### DISCUSSION TITLE
abi3 build with variation in support for M1

### DISCUSSION THREAD

**Discussion by MrBitBucket** [pypa/cibuildwheel#1362] [General]
I made a small extension and based the setup.py on https://github.com/joerick/python-abi3-package-sample where a command class is used to force the tag into a specific form with a check on the python name using an instance of bdist_wheel_abi3. I have my base set as python 3.7 and that appears to work fine for all my builds.

However, I don't know how to change this for the M1 platform where we cannot build for 3.7 as it seems the first build is for 3.8. The Extension has two keywords which need setting, but those cannot be altered by the command class as that needs a distribution argument which comes from deep inside setuptools.

Is there a way to delay construction of the Extension until after I can call the command class get_tag (or its super method).

**Comment by MrBitBucket** [comment]
This seems to work for me https://github.com/MrBitBucket/ci-rl-accel/blob/master/setup.py

**Comment by joerick** [comment]
I don't think it's a problem that the limited ABI goes back on arm64. In limited API mode, it doesn't matter which Python you build with, it's backward-compatible to the tag _and_ forward-compatible.

**Reply by MrBitBucket** [reply]
A colleague pointed me at this which I think he found in the docs 

> Also note that the Limited API is not necessarily stable: compiling with Py_LIMITED_API with Python 3.8 means that the extension will  run with Python 3.12, but it will not necessarily compile with Python 3.12. In particular, parts of the Limited API may be deprecated and  removed, provided that the Stable ABI stays stable.

Which I take to mean that it might be wrong to compile 3.7 limited api with a python 3.8. It might appear to work, but perhaps there's some hidden problem. I'm certainly not an expert at deciphering that quote.

**Reply by henryiii** [reply]
I think (since that's an API), that means it's not guaranteed to compile, but if it does, it should work. In practice so far, it's pretty rare that something is removed from the Limited API. It's probably a good idea to compile with the newest Python and limited ABI setting in your tests, though, just in case.

**Reply by joerick** [reply]
I agree, I'd take it to mean that the ABI is stable, but the API is not. So you if you can compile the limited API, the ABI is stable through the versions that you set with the C macro. I found the source of that quote - https://docs.python.org/3/c-api/stable.html#limited-api-caveats

But, incidentally, It is news to me that the limited API is not guaranteed to be stable. I think that makes the feature request in https://github.com/pypa/cibuildwheel/discussions/1332 more credible - despite the objections around working sdists, people will occasionally want this.
